pkg/http-tunnel/client/client.go

```diff
@@ -0,0 +1,377 @@
+type Client struct {
+	config *clientData
+
+	conn           net.Conn
+	connMu         sync.Mutex
+	httpServer     *http2.Server
+	serverErr      error
+	lastDisconnect time.Time
+	logger         *log.Entry
+}
+
+// ClientConfig is configuration of the Client.
+type ClientConfig struct {
+	// TlsCrtFilePath specifies path to certificate file for tls connection
+	TlsCrtFilePath string
+	// TlsKeyFilePath specifies path to key file for tls connection
+	TlsKeyFilePath string
+	// ServerAddress specifies TCP address of the tunnel server.
+	ServerAddress string
+	// Tunnels specifies the tunnels client requests to be opened on server.
+	Tunnels map[string]*Tunnel
+	// Logger is optional logger. If nil logging is disabled.
+	Logger *log.Entry
+	// AuthToken
+	AuthToken string
+}
+
+type clientData struct {
+	serverAddr      string
+	tlsClientConfig *tls.Config
+	// dialTLS specifies an optional dial function that creates a tls
+	// connection to the server. If dialTLS is nil, tls.Dial is used.
+	dialTLS func(network, addr string, config *tls.Config) (net.Conn, error)
+	// backoff specifies backoff policy on server connection retry. If nil
+	// when dial fails it will not be retried.
+	backoff   Backoff
+	tunnels   map[string]*proto.Tunnel
+	proxy     ProxyFunc
+	logger    *log.Entry
+	idName    string
+	authToken string
+}
+
+// NewClient creates a new unconnected Client based on configuration. Caller
+// must invoke Start() on returned instance in order to connect server.
+func NewClient(config *ClientConfig) (*Client, error) {
+	clientData := &clientData{}
+
+	if config.ServerAddress == "" {
+		return nil, errors.New("provided empty server address")
+	}
+	clientData.serverAddr = config.ServerAddress
+
+	if config.AuthToken == "" {
+		return nil, errors.New("provided empty auth token")
+	}
+	clientData.authToken = config.AuthToken
+
+	if config.TlsKeyFilePath == "" {
+		return nil, errors.New("provided tls key file path empty")
+	}
+	if config.TlsCrtFilePath == "" {
+		return nil, errors.New("provided tls cert file path empty")
+	}
+	tlsconf, err := TlsClientConfig(
+		config.TlsCrtFilePath,
+		config.TlsKeyFilePath,
+		"",
+		config.ServerAddress)
+	if err != nil {
+		log.Error("TLS ", err)
+	}
+	clientData.tlsClientConfig = tlsconf
+
+	logger := config.Logger
+	if logger == nil {
+		l := log.New()
+		l.SetLevel(log.ErrorLevel)
+		logger = log.NewEntry(l)
+	}
+	clientData.logger = logger
+
+	if config.Tunnels == nil {
+		return nil, errors.New("tunnels maping is nil")
+	}
+	clientData.tunnels = MapTunnels(config.Tunnels)
+	clientData.proxy = CreateProxy(config.Tunnels, logger)
+
+	clientData.backoff = ExpBackoff(BackoffConfig{
+		Interval:    DefaultBackoffInterval,
+		Multiplier:  DefaultBackoffMultiplier,
+		MaxInterval: DefaultBackoffMaxInterval,
+		MaxTime:     DefaultBackoffMaxTime,
+	})
+
+	clientData.idName = "" // TODO
+
+	return newClient(clientData)
+}
+
+func newClient(config *clientData) (*Client, error) {
+	if config.serverAddr == "" {
```

Comment:
Does it make sense to check all of these fields below again when we checked them or assigned them in NewClient?